### ISSUE TITLE
Librispeech learns to predict complete training input sentences

### ISSUE THREAD

**Issue by missEnergy** [tensorflow/tensor2tensor#1129]
### Description

For adapting the librispeech tensor2tensor problem to another language, we rewrote the [librispeech](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/data_generators/librispeech.py) problem. First, we customized the data read pipeline because all of our files are locally stored. Secondly, we customized the generation function to match our input format. We also used the option to generate test data during generation. Our updated problem definition is as following:

```python
_TRAIN_DATASET = [
    "downloaded" 
]

def _collect_data(directory, input_ext, transcription_ext):
    """Traverses directory collecting input and target files."""
      # Directory from string to tuple pair of strings
      # key: the filepath to a datafile including the datafile's basename. Example,
      #   if the datafile was "/path/to/datafile.wav" then the key would be
      #   "/path/to/datafile"
      # value: a pair of strings (media_filepath, label)
    data_files = dict()
    for root, _, filenames in os.walk(directory):
        transcripts = [filename for filename in filenames
                       if transcription_ext in filename]
        for transcript in transcripts:
            transcript_path = os.path.join(root, transcript)
            with open(transcript_path, "r") as transcript_file:
                try:
                    for transcript_line in transcript_file:
                        line_contents = transcript_line.strip().split(" ", 1)
                        media_base, label = line_contents
                        key = os.path.join(root, media_base)
                        assert key not in data_files
                        media_name = "%s.%s"%(media_base, input_ext)
                        media_path = os.path.join(root, media_name)
                        data_files[key] = (media_base, media_path, label)
                except:
                    continue
    return data_files

@registry.register_problem()
class CustomProblem(speech_recognition.SpeechRecognitionProblem):

    # Select only the clean data
    TRAIN_DATASETS = _TRAIN_DATASET

    @property
    def num_shards(self):
        return 90
    
    @property
    def num_dev_shards(self):
        return 10

    @property
    def use_subword_tokenizer(self):
        return False
    
    @property
    def is_generate_per_split(self):
        # generate_data will shard the data into TRAIN and EVAL for us.
        return False

    @property
    def dataset_splits(self):
        """Splits of data to produce and number of output shards for each."""
        # 10% evaluation data
        return [{
            "split": problem.DatasetSplit.TRAIN,
            "shards": 90,
        }, {
            "split": problem.DatasetSplit.EVAL,
            "shards": 10,
        }]

    @property
    def use_train_shards_for_dev(self):
        """If true, we only generate training data and hold out shards for dev."""
        return True

    def generator(self, datasets,
                eos_list=None, start_from=0, how_many=0):
        
        del eos_list
        i = 0
        for dir_ in datasets:
            data_dir = os.path.join("..", "CustomProblem", dir_)
            data_files = _collect_data(data_dir, "wav", ".trans.txt")
            data_pairs = data_files.values()

            encoders = self.feature_encoders(None)
            audio_encoder = encoders["waveforms"]
            text_encoder = encoders["targets"]

            for utt_id, media_file, text_data in sorted(data_pairs)[start_from:]:
                if how_many > 0 and i == how_many:
                    return
                i += 1
                wav_data = audio_encoder.encode(media_file)
                try:
                    spk_id, _ = utt_id.split("-")
                    yield {
                        "waveforms": wav_data,
                        "waveform_lens": [len(wav_data)],
                        "targets": text_encoder.encode(text_data),
                        "raw_transcript": [text_data],
                        "utt_id": [utt_id],
                        "spk_id": [spk_id],
                    }
                except:
                    continue

    def generate_data(self, data_dir, tmp_dir, task_id=-1):
        train_paths = self.training_filepaths(
            data_dir, self.num_shards, shuffled=True)
        dev_paths = self.dev_filepaths(data_dir, self.num_dev_shards, shuffled=True)
        
        if self.use_train_shards_for_dev:
            all_paths = train_paths + dev_paths
            generator_utils.generate_files(self.generator(self.TRAIN_DATASETS), all_paths)
            generator_utils.shuffle_dataset(all_paths)

```

The data generation step went fine, both dev and train files were generated. We put it to train on the dataset of about ~10 hours of speech, formatted in librispeech format (so for each 16 kHZ .wav file, a single transcription sentence). After training for about 500k steps, the eval results were:

`INFO:tensorflow:Saving dict for global step 573891: global_step = 573891, 
loss = 0.5653716, 
metrics-custom_problem/targets/accuracy = 0.88907623, 
metrics-custom_problem/targets/accuracy_per_sequence = 0.0,
metrics-custom_problem/targets/accuracy_top5 = 0.9563425, 
metrics-custom_problem/targets/edit_distance = 0.110548995, 
metrics-custom_problem/targets/neg_log_perplexity = -0.60513526`

Note that the accuracy per sequence is 0.0, what does that mean in this case? Other metrics seem high, and the loss low. 

The problem is, whenever we try to infer with this model, we get full proper sentences as output (looks like sentences in our train set), but **not the correct ones**. It seems like the predicted output is just drawn from some pool of (target) sentences. It is not what I would expect even from a poorly trained model (I would expect jibberish or random words). Did we mess up somewhere in the data generation step? Did someone encounter this?

It turned out to be a long issue, thanks in advance!!

**Comment by ErnstTmp** [comment]
I would say that 10 hours of speech is not enough, you need much more.
